### DISCUSSION TITLE
How to create Item Fulfillment from Sales Order with SOAP API ?

### DISCUSSION THREAD

**Discussion by benjy8001** [netsuitephp/netsuite-php#160] [Q&A]
i try to create a Item Fulfillment base on Sales Order, but i got this error :
```
<platformCore:status isSuccess="false"
                     xmlns:platformCore="urn:core_2023_1.platform.webservices.netsuite.com">
    <platformCore:statusDetail type="ERROR">
        <platformCore:code>VALID_LINE_ITEM_REQD</platformCore:code>
        <platformCore:message>You must have at least one valid line item for this transaction.
        </platformCore:message>
    </platformCore:statusDetail>
</platformCore:status>
```

First, i try to add a Item Fulfillment with the createdFrom and Items records :
```
<Body>
    <add>
        <record xsi:type="ItemFulfillment">
            <createdFrom internalId="4069408" type="salesOrder"/>
            <shipStatus>_picked</shipStatus>
            <itemList replaceAll="true">
                <item>
                    <location internalId="124"/>
                    <quantity>1</quantity>
                    <item internalId="48887" type="inventoryItem"/>
                    <orderLine>1</orderLine>
                    <itemIsFulfilled>true</itemIsFulfilled>
                </item>
            </itemList>
        </record>
    </add>
</Body>
```

It fail. Then i try with a initialize request :
```
<Body>
    <initialize>
        <initializeRecord>
            <type>itemFulfillment</type>
            <reference type="salesOrder" internalId="4069408"/>
        </initializeRecord>
    </initialize>
</Body>
```
But it also fail th same way...

Any idea of how to create a Item Fulfillment record with SOAP ?

Thanks !

**Comment by jrebs** [comment]
When I create itemFulfillments, I start with the initialize call, giving the type of salesOrder and the appropriate internalId, just like you have shown at the bottom of the post. This returns me an itemFulfillment record which I manipulate and then send back as an itemFulfillment add() call. Based on the error message at the top of your post, I'd ask you to verify with certainty that the item internalId you're passing is correct, for starters, but if you're actually getting that same error in response to an initialize call (which you state near the end of your post), then that's a little more confusing. I guess I'd check to make sure the item in question is still in your items list and not marked inactive. I think if an item is deactivated, it will cause errors in any calls referring to it. Afraid I can't offer much more than that because it all works fine for me, but good luck.

**Comment by benjy8001** [comment]
HI @jrebs ,

I try many configuration, and i found a way to make it works. When i set a location (and inventory location) to the sales order's items, it fail (error You must have at least one valid line item for this transaction.)

But when i remove the location (and keep t only on sales order), it works ...

I don't know why, but it unlock my situation for now.

**Reply by paulasokolowicz** [reply]
Hi! I am having a similar issue. Can you show the code for that fix?  I don't understand how you could "remove the location".

**Reply by jrebs** [reply]
When you initialize the itemFulfillment based on the salesOrder record, the web services returns you an itemFulfillment object which you will ultimately be passing back to web services as part of an add() call. Before you return the object, you can manipulate it by nulling out properties.

**Reply by benjy8001** [reply]
Hi @paulasokolowicz .

@jrebs , it's not really that, because Netsuite API refuse to initialize an Item Fulfillment.

@paulasokolowicz in original Sales Order, in order lines, i was forced to set the location field to empty value, and then, the API can initialize Item Fulfillment. And then, with the initalize record, i set the location on all items of item fulfillment before save it.

I notice that when i create an Item Fulfillment from Sales Order with the UI, there is a "itemlocation" parameter in the url. But with soap API, we cannot set that value ...

**Reply by CalmEsSimple** [reply]
@benjy8001 thank you! 

i have another question. do you have any example of a retrieve itemfulfullment to get its inventoryDetail?

**Comment by jrebs** [comment]
Glad you found a solution.
